examples/pennylane_run_doubly_stochastic.py

```diff
@@ -0,0 +1,312 @@
+from pennylane.init import strong_ent_layers_uniform
+from pennylane.templates.layers import StronglyEntanglingLayers
+
+num_layers = 2
+num_wires = 2
+eta = 0.01
+steps = 200
+
+dev_GD = qml.device("default.qubit", wires=num_wires, analytic=True)
+dev_SGD = qml.device("default.qubit", wires=num_wires, analytic=False)
+
+##############################################################################
+# We can use ``qml.Hermitian`` to directly specify that we want to measure
+# the expectation value of the matrix :math:`H`:
+
+H = np.array([[8, 4, 0, -6], [4, 0, 4, 0], [0, 4, 8, 0], [-6, 0, 0, 0]])
+
+
+def circuit(params):
+    StronglyEntanglingLayers(*params, wires=[0, 1])
+    return expval(qml.Hermitian(H, wires=[0, 1]))
+
+
+##############################################################################
+# Now, we create three QNodes, each corresponding to a device above,
+# and optimize them using gradient descent via the parameter-shift rule.
+
+qnode_GD = qml.QNode(circuit, dev_GD)
+qnode_SGD = qml.QNode(circuit, dev_SGD)
+
+init_params = strong_ent_layers_uniform(num_layers, num_wires)
+
+# Optimizing using exact gradient descent
+
+cost_GD = []
+params_GD = init_params
+opt = qml.GradientDescentOptimizer(eta)
+
+for _ in range(steps):
+    cost_GD.append(qnode_GD(params_GD))
+    params_GD = opt.step(qnode_GD, params_GD)
+
+# Optimizing using stochastic gradient descent with shots=1
+
+dev_SGD.shots = 1
+cost_SGD1 = []
+params_SGD1 = init_params
+opt = qml.GradientDescentOptimizer(eta)
+
+for _ in range(steps):
+    cost_SGD1.append(qnode_SGD(params_SGD1))
+    params_SGD1 = opt.step(qnode_SGD, params_SGD1)
+
+# Optimizing using stochastic gradient descent with shots=100
+
+dev_SGD.shots = 100
+cost_SGD100 = []
+params_SGD100 = init_params
+opt = qml.GradientDescentOptimizer(eta)
+
+for _ in range(steps):
+    cost_SGD100.append(qnode_SGD(params_SGD100))
+    params_SGD100 = opt.step(qnode_SGD, params_SGD100)
+
+
+##############################################################################
+# We can now plot the cost against optimization step for the three cases.
+
+from matplotlib import pyplot as plt
+
+plt.style.use("seaborn")
+plt.plot(cost_GD[:100], label="Vanilla gradient descent")
+plt.plot(cost_SGD100[:100], label="Stochastic gradient descent, $shots=100$")
+plt.plot(cost_SGD1[:100], ".", label="Stochastic gradient descent, $shots=1$")
+
+# analytic ground state
+min_energy = min(np.linalg.eigvalsh(H))
+plt.hlines(min_energy, 0, 100, linestyles=":", label="Ground-state energy")
+
+plt.ylabel("Cost function value")
+plt.xlabel("Optimization steps")
+plt.legend()
+plt.show()
+
+##############################################################################
+# Using the trained parameters from each optimization strategy, we can
+# evaluate the analytic quantum device:
+
+print("Vanilla gradient descent min energy = ", qnode_GD(params_GD))
+print("Stochastic gradient descent (shots=100) min energy = ", qnode_GD(params_SGD100))

```

Comment:
And here the reason why we want a separate QNode for evaluation --- it is not clear a priori why you evaluate the SGD parameters with the GD QNode.

Reply:
I've updated it to `qnode_analytic` and `qnode_stochastic`, as I realized the QNode is independent of the optimization method, so the previous name didn't really make sense